cupyx/jit/_builtin_funcs.py

```diff
@@ -82,9 +83,34 @@ def call_const(self, env, dtype, size):
         return Data(name, _cuda_types.Ptr(child_type))
 
 
+class AtomicOp(BuiltinFunc):
+
+    def __init__(self, op, dtypes):
+        self._op = op
+        self._name = 'atomic' + op
+        self._dtypes = dtypes
+        super().__init__()
+
+    def call(self, env, array, index, value):
+        array = Data.init(array, env)
+        if not isinstance(array.ctype, (_cuda_types.CArray, _cuda_types.Ptr)):
+            raise TypeError('The first argument must be of array type.')
+        target = _compile._indexing(array, index, env)
+        ctype = target.ctype
+        value = _compile._astype_scalar(value, ctype, 'same_kind', env)
+        name = self._name
+        value = Data.init(value, env)
+        if ctype.dtype.char not in self._dtypes:
+            raise TypeError(f'`{name}` does not support {ctype.dtype} input.')
+        return Data(f'{name}(&{target.code}, {value.code})', ctype)
+
+
 builtin_functions_dict = {
     range: RangeFunc(),
 }
 
 syncthreads = SyncThreads()
 shared_memory = SharedMemory()
+
+# TODO: Add more atomic functions.
+atomic_add = AtomicOp('Add', 'iILefd')
```

Comment:
I think this should be fixed?
```suggestion
atomic_add = AtomicOp('Add', 'iqQefd')
```
Notes:
1. The supported atomics are
```C
int atomicAdd(int* address, int val);
unsigned int atomicAdd(unsigned int* address,
                       unsigned int val);
unsigned long long int atomicAdd(unsigned long long int* address,
                                 unsigned long long int val);
float atomicAdd(float* address, float val);
double atomicAdd(double* address, double val);
__half atomicAdd(__half *address, __half val);
```
and their bit-widths aren't affected by LP64 (linux64) vs LLP64 (win), so we should prefer `q` over `l`.

2. Just a note: by adding `'e'` (for half) here we are dropping the support for CUDA 9.2 (https://github.com/cupy/cupy/issues/5120), as this version was added since 10.0. But I think we should do it.

Reply:
@leofang Thanks for your review!

1. This list of dtypes is used only to check if `atomicAdd` supports the data type of input, and is NOT used to determine the dtype of output. So, I just added `Q` dtype to the list. (973d6e7)
2. I want to backport this PR to v9, so I fixed to raise `RuntimeError` for float16 input for CUDA 9.2. (a419bd5)

Reply:
1. Yes, but in this case input and output happen to have the same data type 🙂 btw, you're right to not add `q`, my mistake!
2. Sounds good!